qa/full-cluster-restart/build.gradle

```diff
@@ -0,0 +1,100 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.elasticsearch.gradle.test.RestIntegTestTask
+import org.elasticsearch.gradle.Version
+
+apply plugin: 'elasticsearch.standalone-test'
+
+// This is a top level task which we will add dependencies to below.
+// It is a single task that can be used to backcompat tests against all versions.
+task bwcTest {
+  description = 'Runs backwards compatibility tests.'
+  group = 'verification'
+}
+
+for (Version version : indexCompatVersions) {
+  String baseName = "v${version}"
+
+  Task oldClusterTest = tasks.create(name: "${baseName}#oldClusterTest", type: RestIntegTestTask) {
+    mustRunAfter(precommit)
+  }
+
+  Object extension = extensions.findByName("${baseName}#oldClusterTestCluster")
+  configure(extensions.findByName("${baseName}#oldClusterTestCluster")) {
+    distribution = 'zip'
+    bwcVersion = version
+    numBwcNodes = 2
+    numNodes = 2
+    clusterName = 'full-cluster-restart'
+    if (version.onOrAfter('5.3.0')) {
+      setting 'http.content_type.required', 'true'
+    }
+  }
+
+  Task oldClusterTestRunner = tasks.getByName("${baseName}#oldClusterTestRunner")
+  oldClusterTestRunner.configure {
+    systemProperty 'tests.is_old_cluster', 'true'
+    systemProperty 'tests.old_cluster_version', version.toString().minus("-SNAPSHOT")
+    systemProperty 'tests.path.repo', new File(buildDir, "cluster/shared/repo")
+  }
+
+  Task upgradedClusterTest = tasks.create(name: "${baseName}#upgradedClusterTest", type: RestIntegTestTask) {
+    dependsOn(oldClusterTestRunner, "${baseName}#oldClusterTestCluster#node0.stop")
+  }
+
+  configure(extensions.findByName("${baseName}#upgradedClusterTestCluster")) {
+    dependsOn oldClusterTestRunner,
+        "${baseName}#oldClusterTestCluster#node0.stop",
+        "${baseName}#oldClusterTestCluster#node1.stop"
+    distribution = 'zip'
+    clusterName = 'full-cluster-restart'
+    numNodes = 2
+    dataDir = { nodeNum -> oldClusterTest.nodes[nodeNum].dataDir }
+  }
+
+  Task upgradedClusterTestRunner = tasks.getByName("${baseName}#upgradedClusterTestRunner")
+  upgradedClusterTestRunner.configure {
+    systemProperty 'tests.is_old_cluster', 'false'
+    systemProperty 'tests.old_cluster_version', version.toString().minus("-SNAPSHOT")
+    systemProperty 'tests.path.repo', new File(buildDir, "cluster/shared/repo")
+  }
+
+  Task versionBwcTest = tasks.create(name: "${baseName}#bwcTest") {
+    dependsOn = [upgradedClusterTest]
+  }
+
+  /* Delay this change because the task we need to modify isn't created until
+   * after projects are evaluated. */
+  gradle.projectsEvaluated {
+    // Disable cleaning the repository so we can test loading a snapshot
+    tasks.getByName("${baseName}#upgradedClusterTestCluster#prepareCluster.cleanShared").enabled = false
```

Comment:
This is kind of sneaky...maybe have a flag on cluster configuration that means "don't create a shared clean task"? Something about "reuse"?